pkg/ingester/ingester_ingest_storage_test.go

```diff
@@ -0,0 +1,164 @@
+		t.Run(fmt.Sprintf("read consistency = %s", readConsistency), func(t *testing.T) {
+			var (
+				cfg     = defaultIngesterTestConfig(t)
+				limits  = defaultLimitsTestConfig()
+				reg     = prometheus.NewRegistry()
+				ctx     = context.Background()
+				series1 = mimirpb.PreallocTimeseries{
+					TimeSeries: &mimirpb.TimeSeries{
+						Labels:  mimirpb.FromLabelsToLabelAdapters(labels.FromStrings(labels.MetricName, metricName)),
+						Samples: []mimirpb.Sample{{TimestampMs: 1000, Value: 10}},
+					},
+				}
+			)
+
+			limits.IngestStorageReadConsistency = readConsistency
+
+			// Create the ingester.
+			overrides, err := validation.NewOverrides(limits, nil)
+			require.NoError(t, err)
+			ingester, kafkaCluster := createTestIngesterWithIngestStorage(t, &cfg, overrides, reg)
+
+			// Mock the Kafka cluster to fail the Fetch operation until we unblock it later in the test.
+			// If read consistency is "eventual" then these failures shouldn't affect queries, but if it's set
+			// to "strong" then a query shouldn't succeed until the Fetch requests succeed.
+			failFetch := atomic.NewBool(true)
+
+			kafkaCluster.ControlKey(int16(kmsg.Fetch), func(kreq kmsg.Request) (kmsg.Response, error, bool) {
+				kafkaCluster.KeepControl()
+
+				if failFetch.Load() {
+					return nil, errors.New("mocked error"), true
+				}
+
+				return nil, nil, false
+			})
+
+			// Start the ingester (after the Kafka cluster has been mocked).
+			require.NoError(t, services.StartAndAwaitRunning(ctx, ingester))
+			defer t.Cleanup(func() {
+				require.NoError(t, services.StopAndAwaitTerminated(ctx, ingester))
+			})
+
+			// Wait until the ingester is healthy.
+			test.Poll(t, 1*time.Second, 1, func() interface{} {
+				return ingester.lifecycler.HealthyInstancesCount()
+			})
+
+			// Create a Kafka writer and then write a series.
+			writer := ingest.NewWriter(cfg.IngestStorageConfig.KafkaConfig, log.NewNopLogger(), nil)
+			require.NoError(t, services.StartAndAwaitRunning(ctx, writer))
+			defer t.Cleanup(func() {
+				require.NoError(t, services.StopAndAwaitTerminated(ctx, writer))
+			})
+
+			partitionID, err := ingest.IngesterPartition(cfg.IngesterRing.InstanceID)
+			require.NoError(t, err)
+			require.NoError(t, writer.WriteSync(ctx, partitionID, userID, &mimirpb.WriteRequest{Timeseries: []mimirpb.PreallocTimeseries{series1}, Source: mimirpb.API}))
+
+			// Run a query in a separate goroutine and collect the result.
+			var (
+				queryRes model.Matrix
+				queryWg  = sync.WaitGroup{}
+			)
+
+			queryWg.Add(1)
+			go func() {
+				defer queryWg.Done()
+
+				// Ensure the query will eventually terminate.
+				queryCtx, cancel := context.WithTimeout(user.InjectOrgID(ctx, userID), 5*time.Second)
+				defer cancel()
+
+				queryRes, _, err = runTestQuery(queryCtx, t, ingester, labels.MatchEqual, labels.MetricName, metricName)
+				require.NoError(t, err)
+			}()
+
+			// Wait some time and then unblock the Fetch.
+			time.Sleep(time.Second)
```

Comment:
is it possible to do this without a Sleep? Sleeps are usually flaky it might end up just not doing what we need it to (also slows down the test).

If my understanding is correct we want to wait so that the Fetch fails a couple of times before we unblock it. So we want to test resiliency during startup. Can we just close a channel in the ControlKey once we get the first error? Or perhaps count the number of invocations and unblock the fetch only after the first few invocations have failed?

Reply:
_I start from the assumption that "every sleep in a test is bad" is not the right approach (similarly to "any use of panic() in the code is bad"... well it depends). There are times when a sleep is good enough to get what you want, it's reliable when the sleep time is large enough, and sometimes it's 10x times easier than alternatives. That being said..._

We want to unblock the Fetch **while** the query is waiting on `Ingester.enforceReadConsistency()`. If we do as you say, that is failing the first few Fetch requests, well it's not much different than a sleep because Fetch (or ListOffsets, that's the other thing we could fail) are issued in a loop which is completely independent from the query path.

A better approach may be unblock the Fetch right before calling `runTestQuery()`. There's still no guarantee that we would get execution in the other we want (at the end, the Fetch may run before the actual query is waiting for consistency) so to make it further reliable we may have to also add a short sleep (oh, welcome back sleep!), or we could ignore the problem assuming that in practice it will be very unlikely to happen.

Reply:
> A better approach may be unblock the Fetch right before calling runTestQuery()

Done here: https://github.com/grafana/mimir/pull/7030/commits/79d0a1924660315be0dab85b42a7a0457597b788. Even if there's no strong guarantee that the query will wait for read consistency before we unblock the fetch, I think that in practice it's nearly impossible to get the execution in the wrong order because of the ListOffsets fetched in a loop.